### ISSUE TITLE
[torch.compile] Add `explain` as a backend

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#102053]
[`torch._dynamo.explain`](https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/_dynamo/eval_frame.py#L562) is a useful API for PT2 users. Users can find the number of graphs and graph breaks using this API, giving them actionable messages. One of the pain points is - its a different API than torch.compile. @Chillee  proposed to add it as a backend, simplifying the UX issue. 

~~~
import torch
import torch._dynamo

def fn3(x):
    x = torch.sin(x)
    torch._dynamo.graph_break()
    x = torch.sin(x)
    return x



def fn2(x):
    x = torch.cos(x)
    x = fn3(x)
    x = torch.cos(x)
    return x


def fn1(x):
    x = torch.tan(x)
    torch._dynamo.graph_break()
    x = fn2(x)
    x = torch.tan(x)
    return x

def fn(x):
    x = torch.sigmoid(x)
    x = fn1(x)
    x = torch.sigmoid(x)
    return x


# outs = torch._dynamo.explain(fn, torch.randn(4))
explanation, out_guards, graphs, ops_per_graph, break_reasons, explanation_verbose = torch._dynamo.explain(fn, torch.randn(4))
print(explanation_verbose)
for graph in graphs:
    print(graph.graph.print_tabular())

~~~    

### Short term task

Introduce a new backend - `explain` such that `torch.compile(fn, backend="explain")` gives morally same output as torch._dynamo.explain. We can add the `explain` backend in this file - [torch/_dynamo/backends/debugging.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/backends/debugging.py)

As a start, we can focus on graph breaks and explanation string. We would want the backend to be configurable so that we can expand to other stuff.

### Improving the output

* More details around graph breaks. For example, a graph break deep in the call stack can lead to many graph breaks in the model. The output should explain this part well, and potentially rank the graph breaks.
* Recompiles - And reason of recompile

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh

**Comment by alimoezzi** [comment]
Hi, 
I have [this](https://github.com/pytorch/pytorch/pull/103249) implementation for explain as backend.
For now it by default logs graph breaks as follows:
```
Graph Break
Break reason:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 37, in fn
    x = fn1(x)
  File "~/pytorch/main3.py", line 30, in fn1
    torch._dynamo.graph_break()

Graph Break
Break reason:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 30, in fn1
    torch._dynamo.graph_break()

Graph Break
Break reason:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 23, in fn2
    x = fn3(x)
  File "~/pytorch/main3.py", line 15, in fn3
    torch._dynamo.graph_break()

Graph Break
Break reason:
 call_function graph_break in skip_files ~/pytorch/torch/_dynamo/__init__.py
  File "~/pytorch/main3.py", line 15, in fn3
    torch._dynamo.graph_break()
```
and by configuration it can be enabled to log graphs as well.
I used the example here as well for writing the test for explain backend.
Let me know if it needs any change.
